server/runtime.go

```diff
@@ -105,16 +98,91 @@ func NewRuntime(logger *zap.Logger, multiLogger *zap.Logger, db *sql.DB, config
 		return nil, err
 	}
 
-	multiLogger.Info("Evaluating modules", zap.Int("count", len(modules)), zap.Strings("modules", modules))
-	if err = r.loadModules(lua.LuaLDir, modules); err != nil {
+	// Initialize a one-off runtime to ensure startup code runs and modules are valid.
+	vm := lua.NewState(lua.Options{
+		CallStackSize:       1024,
+		RegistrySize:        1024,
+		SkipOpenLibs:        true,
+		IncludeGoStackTrace: true,
+	})
+	for name, lib := range stdLibs {
+		vm.Push(vm.NewFunction(lib))
+		vm.Push(lua.LString(name))
+		vm.Call(1, 0)
+	}
+	nakamaModule := NewNakamaModule(logger, db, vm, notificationService, true)
+	vm.PreloadModule("nakama", nakamaModule.Loader)
+	r := &Runtime{
+		logger: logger,
+		vm:     vm,
+		luaEnv: ConvertMap(vm, config.Environment),
+	}
+	moduleStrings := make([]string, len(modules))
+	for i, module := range modules {
+		moduleStrings[i] = module.path
+	}
+	multiLogger.Info("Evaluating modules", zap.Int("count", len(moduleStrings)), zap.Strings("modules", moduleStrings))
+	if err = r.loadModules(modules); err != nil {
 		return nil, err
 	}
 	multiLogger.Info("Modules loaded")
+	r.Stop()
+
+	return &RuntimePool{
+		pool: &sync.Pool{
+			New: func() interface{} {
+				vm := lua.NewState(lua.Options{
+					CallStackSize:       1024,
+					RegistrySize:        1024,
+					SkipOpenLibs:        true,
+					IncludeGoStackTrace: true,
+				})
+
+				for name, lib := range stdLibs {
+					vm.Push(vm.NewFunction(lib))
+					vm.Push(lua.LString(name))
+					vm.Call(1, 0)
+				}
+
+				nakamaModule := NewNakamaModule(logger, db, vm, notificationService, false)
+				vm.PreloadModule("nakama", nakamaModule.Loader)
+
+				r := &Runtime{
+					logger: logger,
+					vm:     vm,
+					luaEnv: ConvertMap(vm, config.Environment),
+				}
+
+				if err = r.loadModules(modules); err != nil {
+					multiLogger.Fatal("Failed initializing runtime modules", zap.Error(err))
+				}
+				return r
+
+				// TODO find a way to run r.Stop() when the pool discards this runtime.
```

Comment:
Is this still a valid TODO? Do we need to do something about this?

Reply:
@mofirouz It is valid, but not critical because it only explicitly cleans up temp files if any. We don't expose any Lua stdlib functions that access temp files.

There's no way to hook into `sync.Pool`'s eviction process so the only way to fully solve this is to use a different pool implementation.